### DISCUSSION TITLE
Gridding tutorial boundary gotchas

### DISCUSSION THREAD

**Discussion by joshuacortez** [thinkingmachines/geowrangler#140] [Usability Testing Q&A]
In the [gridding tutorial](https://geowrangler.web.app/tutorial.grids.html), the bounding box was directly computed from the epsg:4326 admin bounds

`aoi_total_bounds = region3_gdf.total_bounds`

These bounds were then passed to `SquareGridGenerator`

`grid_generator1k = grids.SquareGridGenerator(1_000, boundary=aoi_total_bounds)`

In my own use case, I tried to make 100x100m grid cells across the entire Indonesian coastline. In the same way as the tutorial, I computed the bounding box on the epsg:4326 geodataframe, and passed them to `SquareGridGenerator`.

However I found that there was a significant number of missing grid cells on the northwestern side of Indonesia. 
<img width="567" alt="uncaptured portion" src="https://user-images.githubusercontent.com/7113605/186683316-f92eac95-4643-4caf-a632-b640fe637101.PNG">

There are supposed to be tiles along Aceh's coast. The image below shows what should have been there. (tiles are smaller since they're from a separate run)
<img width="441" alt="missing coastline grids" src="https://user-images.githubusercontent.com/7113605/186683574-e1cdcbca-eaa5-417d-88ac-bed7e4dcd200.PNG">

After some investigation, I found that it had to do with the bounding box. **I should have computed the bounding box on the projected admin boundaries, not the epsg:4326 admin boundaries**. Even if the `proj_crs` is specified in `SquareGridGenerator`, we’ll still have this error because the bounding box (computed from epsg:4326) was already distorted to begin with. 

I resolved the issue by generating the bounding box from the projected admin bounds instead of the epsg:4326 admin bounds, and used `SquareGridBoundary` as follows

`boundary = grids.SquareGridBoundary(*proj_admin_bounds_bbox)` 

After passing this `boundary` to `SquareGridGenerator`, tiles now show up along Aceh's coast.

So technically this isn't a bug with geowrangler itself but it's definitely something to watch out for. Hopefully other people won't make the same mistake. I was thinking if we should be explicit about this in the tutorial. We could: 
- Use the projected bbox by default in the tutorial, so people who follow along will use it as best practice. (leaning on this one)
- Or we could keep the same flow but with a warning to use the projected bbox especially for bigger areas.

Let me know what you think! :)

**Comment by butchtm** [comment]
@joshuacortez  -- just clarifying:

in the `generate_grid`  method:
 
*  if you don't pass in a `boundary`  arg when the `SquareGridGenerator` was created -- the boundary is created from the reprojected total bounds of the gdf you pass in the `generate_grid` method. -- _this is not the use case for your example above_

* if a `SquareGridBoundary` as the `boundary`  arg when `SquareGridGenerator`  was created, it is assumed the user already did the reprojecting and computation of the reprojected boundaries. -- _this is what you did to fix the problem, right?_
 
* if an array (`[minx,miny,maxx,maxy]`) (e.g. `aoi.total_bounds`) is passed, the boundaries are reprojected and recomputed -- **is this not the behavior you want? (If it didn't then this might be a bug?)**


```
   ...
    reprojected_gdf = gdf.to_crs(self.grid_projection)
    if self.boundary is None:
        boundary = SquareGridBoundary(*reprojected_gdf.total_bounds)
    elif isinstance(self.boundary, SquareGridBoundary):
        boundary = self.boundary
    else:
        transformer = Transformer.from_crs(gdf.crs, reprojected_gdf.crs, always_xy=True)
        x_min, y_min = transformer.transform(self.boundary[0], self.boundary[1])
        x_max, y_max = transformer.transform(self.boundary[2], self.boundary[3])
        boundary = SquareGridBoundary(x_min, y_min, x_max, y_max)

```

**Reply by joshuacortez** [reply]
> if you don't pass in a boundary arg when the SquareGridGenerator was created -- the boundary is created from the reprojected total bounds of the gdf you pass in the generate_grid method. -- this is not the use case for your example above


Yes that's right

> if a SquareGridBoundary as the boundary arg when SquareGridGenerator was created, it is assumed the user already did the reprojecting and computation of the reprojected boundaries. -- this is what you did to fix the problem, right?


Yes that's what I did

> if an array ([minx,miny,maxx,maxy]) (e.g. aoi.total_bounds) is passed, the boundaries are reprojected and recomputed -- is this not the behavior you want? (If it didn't then this might be a bug?)


Yeah Ideally this is the behavior I'd like to have, but I think there's a sneaky thing with projections. 😅  I think it boils down to: the projection of the epsg:4326 admin bounds bbox != bbox of projected CRS admin bounds. For small areas the difference might be small, but for large areas like an entire country, there's enough distortion to have a noticeable difference.

Example for PHL Admin Bounds (level 3):
- Bbox of epsg:4326: (114.27789225,   4.5872945 , 126.6049656 ,  21.12188548)
- Bbox of epsg:3123 (-236221.24150117,  507465.65086805, 1119703.39754332,
       2336815.97643074)

Ideally, if we project the bbox of epsg:4326, it should match the bbox of epsg:3123, but actually it turns out it's off by as much as 10km
(-247776.82233019953, 510820.26522794546, 1082817.5512844631, 2346808.314776758)

Similarly, if we reproject the bbox of epsg:3123 back to epsg:4326, it can already be off even at the first decimal place
(114.38160217994552, 4.558134851818131, 126.95462932409919, 21.020038098627538)

So yeah it's tricky 😅

**Reply by joshuacortez** [reply]
I think there won't be distortions if the bounding box is represented as 4 pairs of coordinates instead of minx,miny,maxx,maxy. I.e. something like ((minx, miny), (minx, maxy), (maxx, miny), (maxx, maxy)). You could project these 4 separately and get the projected minx and proj miny after

Before what we were doing was just getting the projections of the bottom left and upper right corners. 

> transformer = Transformer.from_crs(gdf.crs, reprojected_gdf.crs, always_xy=True)
> x_min, y_min = transformer.transform(self.boundary[0], self.boundary[1])
> x_max, y_max = transformer.transform(self.boundary[2], self.boundary[3])
> boundary = SquareGridBoundary(x_min, y_min, x_max, y_max)

But it might be more accurate to get the projections of 4 corners

That could be another approach? I didn't think of this yesterday but if we do this then we can keep the same API. I'll try testing it

**Reply by butchtm** [reply]
asking @jtmiclat if there's an alternative way to get better boundaries  here...

**Reply by joshuacortez** [reply]
I tried to get a bounding box after doing a projection on the 4 corners like this

```
transformer = Transformer.from_crs(gdf.crs, reprojected_gdf.crs, always_xy=True)
# lower left corner
x1, y1 = transformer.transform(self.boundary[0], self.boundary[1])  
# upper right corner
x2, y2 = transformer.transform(self.boundary[2], self.boundary[3])  
# lower right corner
x3, y3 = transformer.transform(self.boundary[2], self.boundary[1])  
# upper left corner
x4, y4 = transformer.transform(self.boundary[0], self.boundary[3]) 

x_min = min([x1, x2, x3, x4])
x_max = max([x1, x2, x3, x4])

y_min = min([y1, y2, y3, y4])
y_max = max([y1, y2, y3, y4])
```

It yielded this bounding box: (-247776.82233019953, 509703.88359714346, 1122742.1893233713, 2351381.9997901437)

It still doesn't match the actual bounding box of the projected admin bounds, it's off by 7.8km on average. Although, it is half of the error using the original approach of the upper right and lower left corners (15km). But still there's a distortion.

Looks like we need to project first before getting the bounding box to avoid distortions? I'll do another experiment

**Reply by jtmiclat** [reply]
i was looking at getting the bounds at 4 corners ala what Joshua has done but there are still edge cases for very large areas and small grids especially the areas that intersects with the equator.


I've faced some issues with reprojections before and already added a buffer to account for small rounding error differences but this doesn't really solve all reprojection problems 

https://github.com/thinkingmachines/geowrangler/blob/31b47d58499d6260bdbfd46783c5ca896af03d51/geowrangler/grids.py#L44-L52

Some possible solutions:
1. Remove reprojection of bounds within geowrangler and ask the users to give bounds in EPSG:3857 if they plan to use custom boundaries
2. Allow grid cell going outside bounds to have a negative index. The main reason we have bounds is so they have consistent x and y indexes throughout different grid generation. It made sense that these indexes range from 0-x_max_gridsize or 0-y_max_gridsize and everything out side this index is removed. If we allow there to be negative indexes, grid cells generated outside the bounds should still exist. 

If we go with 2, we might not need boundary of the aoi but instead ask for "origin" (0, 0) of the index.

**Reply by joshuacortez** [reply]
I tried projecting the entire admin bound bounding box but it has the same results as my 2nd experiment. It also yielded this bounding box (-247776.82233019953, 509703.88359714346, 1122742.1893233713, 2351381.9997901437)

```
from shapely.geometry import box

bbox_projected = (
    gpd.GeoSeries(box(*admin_bounds_bbox), crs="epsg:4326").to_crs(PROJ_CRS).bounds
)

x_min, y_min, x_max, y_max = (
    bbox_projected["minx"][0],
    bbox_projected["miny"][0],
    bbox_projected["maxx"][0],
    bbox_projected["maxy"][0],
)
x_min, y_min, x_max, y_max
```

**Reply by joshuacortez** [reply]
@jtmiclat Related to the point about removing reprojection of bounds, what do you think about removing reprojections from geowrangler gridding entirely? The user can be expected to pass in both bounds and AOIs that are already reprojected. It could be the same as the behavior when using geopandas `.area` and `.distance` where geopandas gives this warning

UserWarning: Geometry is in a geographic CRS. Results from 'area' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

The workflow of projecting first before doing operations that require measurements in meters can be aligned to what geopandas users are already familiar with. With this we don’t need negative indices too right?

**Reply by jtmiclat** [reply]
If our only target users are GIS people i would be more willing to remove auto reprojection, but if we are also targeting Data scientists without Geospatial experience the concept of projection might be challenging. This edge case only happens when we give custom bounds which is more of advanced use case so removing auto reprojection is really catered to advance users.  

For advanced users, we might just add a caveat, additional instructions, or a warning to reproject when passing the bounds?  Something like
```
>>> grid_generator1k = grids.SquareGridGenerator(1_000, boundary=aoi_total_bounds)
>>> grid_generator1k.generate_grid(region3_gdf)
Warning: The bounds of aoi after reprojection is outside the boundary after reprojection. There might be missing grid cells due to this. Trying adding a buffer to the boundary or reproject aoi before getting the boundary. See geowrangler.thinkingmachin.es/grids#repojectioning-before-gridding 
```

**Reply by joshuacortez** [reply]
This makes a lot of sense. With this usage, the `aoi_total_bounds` are reprojected, while the aois are not reprojected, and the user passes in the crs string?

**Reply by jtmiclat** [reply]
what i imagined   
```
>>> region3_gdf = load_gdf()
>>> grid_generator1k = grids.SquareGridGenerator(1_000, boundary=region3.total_bounds)
>>> grid_generator1k.generate_grid(region3_gdf.subset(1))
Warning: The bounds of aoi after reprojection is outside the boundary after reprojection. There might be missing grid cells due to this. Trying adding a buffer to the boundary or reproject aoi before getting the boundary. See geowrangler.thinkingmachin.es/grids#repojectioning-before-gridding 



... default projection is epsg:3857
>>> reprojected_gdf = region3_gdf.to_crs("EPSG:3857")
>>> grid_generator1k = grids.SquareGridGenerator(1_000, boundary=reprojected_gdf.total_bounds)
>>> grid_generator1k.generate_grid(reprojected_gdf.subset(1))



... if with custom crs
>>> reprojected_gdf = region3_gdf.to_crs(custom_crs)
>>> grid_generator1k = grids.SquareGridGenerator(1_000, boundary=reprojected_gdf.total_bounds, grid_projection=custom_crs)
>>> grid_generator1k.generate_grid(reprojected_gdf.subset(1))

**Reply by joshuacortez** [reply]
Ok got it, so no need to pass in `custom_crs` to `generate_grid`

**Reply by joshuacortez** [reply]
Just to ask, did we get to implement this? :D

**Reply by jtmiclat** [reply]
We haven't! But we created an issue to track it: https://github.com/thinkingmachines/geowrangler/issues/147

**Reply by joshuacortez** [reply]
cool cool cool

**Comment by joshuacortez** [comment]
Is addressed by #253 

1. `SquareGridGenerator` and `FastSquareGridGenerator` should provide a warning if the boundary does not fully enclose the AOI
2. Added markdown text documentation in `tutorial.grids.ipynb` that explains using boundary of reprojected AOI instead of original AOI
